README.md

```diff
@@ -1 +1,471 @@
+compatible with the holochain conductors and host these anywhere.
+
+The main concern is that the `random` op is opaque from the caller's point of
+view and this is where CloudFlare could collude with (or be hacked by) an
+attacker to "randomly" only return malicious nodes.
+
+The additional ops `get` and `list` are provided so that extra paranoid agents
+can implement their own randomness at the expense of additional network hops.
+
+## API
+
+### POST method
+
+All API requests use the HTTP POST method.
+
+This is for both gets and sets.
+
+This is so that we can use messagepack binary data as-is for all requests and
+responses with no 'extra steps' like handling base64 encoding/decoding and URL
+parsing just to work with binary data.
+
+### GET ping
+
+There is _one_ GET endpoint, used for debugging, testing and health checks.
+
+All GET requests to the bootstrap service will receive the response string `OK`
+encoded as UTF-8 text data and the `200` status code.
+
+The GET endpoint behaves differently to all other methods in that it is not
+serialized, is not binary data, has no op headers and cannot interact with the
+kv service at all.
+
+### MessagePack serialization
+
+All requests and responses are serialized as binary MessagePack data in the
+body of the request/response.
+
+The op header (see below) defines which operation the messagepack payload in a
+request is dispatched to and what kind of response will be returned.
+
+This is the same serialization format that holochain itself uses for wire
+messages on the network and for compatibility with wasm for validation logic.
+
+### Ed25519/NaCl crytography
+
+All cryptographic logic is handled as per libsodium using the Ed25519 curve.
+
+This is the same cryptography as holochain itself which means the signatures and
+validation used by the bootstrap service are the same as those used by the agent
+DHT by conductors.
+
+This implementation uses [tweetnacl](https://www.npmjs.com/package/tweetnacl).
+
+### Headers
+
+The `Content-Type` header for all POST requests must be `application/octet` to
+signify to the server that the body of the request is a binary payload.
+
+The action to be performed is set by the `X-Op` header in the POST request.
+
+The possible values are:
+
+- `put`: store signed agent info
+- `list`: list all stored agent info
+- `get`: retrive a single agent info
+- `random`: retrieve up to N random agents (hybrid of list and get)
+
+## Data structures
+
+### Agent info
+
+`AgentInfoSigned` is the main data structure.
+
+It is the same structure defined by `kitsune_p2p` in the Rust codebase for the
+conductor, but ported to typescript for validation here.
+
+It looks like this on the wire:
+
+```typescript
+{
+ signature: Uint8Array,
+ agent: Uint8Array,
+ agent_info: Uint8Array,
+}
+```
+
+Where the `agent_info` is messagepack serialized binary data that MUST be valid
+for the `agent` public key and `signature` bytes according to libsodium.
```

Comment:
```suggestion
for the 32 byte `agent` public key and 64 `signature` bytes according to libsodium.
```

Reply:
Actually my edit isn't quite correct... do we want to mention that the agent key is 36 bytes, but we just use the first 32 as the agent pubkey?

Reply:
i'll do a followup with 39 bytes for the full canonical bytes